src/tests/ftest/util/server_utils.py

```diff
@@ -49,6 +53,189 @@ class ServerFailed(Exception):
     """Server didn't start/stop properly."""
 
 
+class ServerCommand(CommandWithParameters):
+    """Defines a object representing a server command."""
+
+    def __init__(self, hosts, path):
+        """Create a server Command object"""
+        super(ServerCommand, self).__init__(path, "daos_server")
+
+        self.hosts = hosts
+        self.process = None
+        self.hostfile = None
+
+        self.request = BasicParameter("{}")
+        self.action = BasicParameter("{}")
+        self.targets = FormattedParameter("-t {}")
+        self.config = FormattedParameter("-o {}")
+        self.port = FormattedParameter("-p {}")
+        self.storage = FormattedParameter("-s {}")
+        self.modules = FormattedParameter("-m {}")
+        self.xshelpernr = FormattedParameter("-x {}")
+        self.firstcore = FormattedParameter("-f {}")
+        self.group = FormattedParameter("-g {}")
+        self.attach = FormattedParameter("-a {}")
+        self.sock_dir = FormattedParameter("-d {}")
+        self.insecure = FormattedParameter("-i", None)
+
+    def prepare(self, path, slots):
+        """Prepare the hosts before starting daos server.
+
+        Args:
+            path (str): location to write the hostfile
+            slots (int): slots per host to use in the hostfile
+
+        Raises:
+            ServerFailed: if there is any errors preparing the hosts
+
+        """
+        # Kill any doas servers running on the hosts
+        kill_server(self.hosts)
+
+        # Clean up any files that exist on the hosts
+        clean_server(self.hosts)
+
+        # Ensure the environment for the daos server on each host
+        run_agent(path, self.hosts)
+
+        # Create the hostfile
+        self.hostfile = write_host_file(self.hosts, path, slots)
+
+    def update_configuration(self, basepath):
+        """Update the config parameter with a yaml file.
+
+        Args:
+            basepath (str): DAOS install basepath
+        """
+        self.config.update(create_server_yaml(basepath), "server.config")
+
+    def get_param_names(self):
+        """Get a sorted list of daos_server command parameter names."""
+        names = self.get_attribute_names(FormattedParameter)
+        names.extend(["request", "action"])
+        return names
```

Comment:
This looks identical to:
https://github.com/daos-stack/daos/blob/dd5111672e3983f52f42176ded65f0fee2c2d3be/src/tests/ftest/util/dmg_utils.py#L63-L67 (including the copy'n'pasto).  Is there any opportunity to look at the class hierarchy here and [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) this out?  Just seems odd that the two children of this class are overriding the parent in exactly the same way.

Reply:
Agreed, the code duplicity seems unnecessary... this is mainly an attempt to get the command parameters in a certain order and since the dmg command and server command are similar in how the parameters are ordered both have this overriden method. I think that until we have another instance of a class that has to override this function the same way, we should keep the ObjectWithParameters.get_param_names() as is, but I'll think of a way to get away from duplicating this function.

Reply:
Something like this should do the trick:

```
class DaosCommand(CommandWithParameters):

    def __init__(self, path, command):
        super(ServerCommand, self).__init__(os.path.join(path, command))
        self.request = BasicParameter("{}")
        self.action = BasicParameter("{}")

    def get_param_names(self): 
        """Get a sorted list of daos_server command parameter names.""" 
        names = self.get_attribute_names(FormattedParameter) 
        names.extend(["request", "action"]) 
        return names

class ServerCommand(DaosCommand): ...

class DmgCommand(DaosCommand): ...
```